### PR TITLE
Collect and Store Performance Data and Store in Backend Logging API

### LINKED ISSUE #43: Advanced Algorithm Analysis Dashboard with ML Insights
## 🎯 Feature Overview
**Feature Title**  
Advanced Algorithm Analysis Dashboard with ML Insights

**Feature Category**  
- [x] 📊 Advanced Analytics
- [x] 🤖 Machine Learning
- [x] 🎓 Educational Content
- [x] 🔬 Research Tools

## 🔍 Problem Statement
**Is your feature request related to a problem? Please describe.**  
Current performance metrics are basic (time, swaps, comparisons) and don't provide predictive insights, pattern recognition, or intelligent recommendations. Users can't understand long-term performance trends, algorithm suitability for different data patterns, or receive AI-powered optimization suggestions.

**User Story**  
As a researcher or advanced student studying algorithm performance, I want intelligent analysis and predictions about algorithm behavior so that I can make data-driven decisions about algorithm selection and understand complex performance patterns across different scenarios.

## 💡 Proposed Solution
**Describe the solution you'd like**  
Implement an advanced analytics dashboard powered by machine learning that provides:
- **Performance Prediction**: ML models predicting algorithm performance for different input sizes/patterns
- **Pattern Recognition**: Identify optimal algorithms for specific data characteristics
- **Trend Analysis**: Historical performance tracking with anomaly detection
- **Smart Recommendations**: AI-powered algorithm selection based on use case
- **Performance Clustering**: Group similar algorithm runs and identify patterns
- **Predictive Scaling**: Forecast performance at larger scales

**Key Features/Requirements:**
- [ ] ML model training on algorithm performance data
- [ ] Real-time performance prediction engine
- [ ] Interactive data visualization dashboard
- [ ] Pattern recognition for data characteristics
- [ ] Historical performance trend analysis
- [ ] Anomaly detection for performance regressions
- [ ] Smart algorithm recommendation system
- [ ] Performance forecasting for large datasets
- [ ] Export ML insights and reports
- [ ] A/B testing framework for algorithm comparison

**Acceptance Criteria:**
- [ ] ML models achieve >85% accuracy in performance prediction
- [ ] Dashboard loads and updates in real-time without lag
- [ ] Pattern recognition correctly identifies data characteristics
- [ ] Recommendations are relevant and actionable
- [ ] Historical data is stored and queryable efficiently
- [ ] Works across all supported algorithms and array sizes
- [ ] Mobile-responsive dashboard design

## 🔄 Alternative Solutions
**Describe alternatives you've considered**  
- Static performance charts lack predictive capabilities
- Manual algorithm selection requires deep expertise
- External analytics tools don't integrate with sorting context

**Why is this the best approach?**
ML-powered insights provide intelligent guidance that adapts to user behavior and data patterns, transforming SortVision from a visualization tool into an intelligent algorithm analysis platform.

## 🎨 Design & Implementation Ideas
**Technical Considerations:**
- **Frontend**: React with advanced charting libraries (D3.js, Recharts)
- **ML Framework**: TensorFlow.js for client-side inference
- **Data Storage**: IndexedDB for local performance history
- **Analytics**: Web Workers for ML computations
- **Visualization**: Canvas-based high-performance charts

**Advanced ML Implementation:**
```javascript
// Performance prediction ML model
class AlgorithmPerformancePredictor {
  constructor() {
    this.model = null;
    this.isTraining = false;
    this.trainingData = [];
  }
  
  async initializeModel() {
    // Initialize TensorFlow.js model for performance prediction
    this.model = tf.sequential({
      layers: [
        tf.layers.dense({inputShape: [6], units: 64, activation: 'relu'}),
        tf.layers.dropout({rate: 0.2}),
        tf.layers.dense({units: 32, activation: 'relu'}),
        tf.layers.dense({units: 1, activation: 'linear'})
      ]
    });
    
    this.model.compile({
      optimizer: 'adam',
      loss: 'meanSquaredError',
      metrics: ['meanAbsoluteError']
    });
  }
  
  async predictPerformance(algorithm, arraySize, dataPattern) {
    const features = this.extractFeatures(algorithm, arraySize, dataPattern);
    const prediction = this.model.predict(tf.tensor2d([features]));
    return prediction.dataSync()[0];
  }
  
  extractFeatures(algorithm, arraySize, dataPattern) {
    return [
      this.encodeAlgorithm(algorithm),
      Math.log(arraySize),
      this.encodeDataPattern(dataPattern),
      arraySize / 1000,
      this.getAlgorithmComplexity(algorithm),
      this.getDataPatternComplexity(dataPattern)
    ];
  }
}

// Pattern recognition system
class DataPatternAnalyzer {
  analyzeArray(array) {
    return {
      sortedness: this.calculateSortedness(array),
      duplicateRatio: this.calculateDuplicateRatio(array),
      variance: this.calculateVariance(array),
      entropy: this.calculateEntropy(array),
      clusteriness: this.calculateClusteriness(array)
    };
  }
  
  recommendOptimalAlgorithm(pattern, constraints) {
    const features = [
      pattern.sortedness,
      pattern.duplicateRatio,
      pattern.variance,
      constraints.timeLimit,
      constraints.memoryLimit
    ];
    
    return this.algorithmRecommendationModel.predict(features);
  }
}

// Performance anomaly detection
class AnomalyDetector {
  detectPerformanceAnomaly(currentMetrics, historicalData) {
    const normalizedCurrent = this.normalizeMetrics(currentMetrics);
    const historicalMean = this.calculateMean(historicalData);
    const historicalStd = this.calculateStandardDeviation(historicalData);
    
    const zScore = Math.abs((normalizedCurrent - historicalMean) / historicalStd);
    return {
      isAnomaly: zScore > 2.5,
      severity: this.calculateSeverity(zScore),
      explanation: this.explainAnomaly(currentMetrics, historicalMean)
    };
  }
}
```

**Dashboard Components:**
- **Performance Predictor**: Input parameters, get ML predictions
- **Algorithm Recommender**: Smart suggestions based on data analysis
- **Trend Analyzer**: Historical performance visualization
- **Pattern Insights**: Data characteristic analysis
- **Anomaly Monitor**: Real-time performance anomaly detection

## 📊 Impact Assessment
**Priority/Importance**  
- **Priority**: 
  - [x] 🟡 Medium (Advanced research feature)

- **Impact**: 
  - [x] 🌟 Critical (Revolutionary for algorithm research)

**Target Audience:**
- [x] 🎓 CS researchers and PhD students
- [x] 👩‍🏫 Advanced CS educators
- [x] 🔬 Algorithm optimization specialists
- [x] 👨‍💻 Performance engineers
- [x] 📊 Data scientists working with algorithms

## 🎯 SSOC Season 4 Information
**Project Status:**
- [x] 🎯 Open for SSOC Season 4 Contributors
- [ ] 🔒 Reserved for Specific Contributor
- [ ] ⏳ In Progress
- [ ] ✅ Completed

**Difficulty Level:**
- [ ] 🟢 Beginner (20 points) - Good first issue, basic React/JS
- [ ] 🟡 Intermediate (30 points) - Moderate complexity, React/state management
- [x] 🔴 Advanced (40 points) - Complex implementation, advanced concepts

**Estimated Time:** 3-4 weeks

**Skills Required:**
- [x] Advanced JavaScript/TypeScript
- [x] Machine Learning (TensorFlow.js)
- [x] Advanced React with performance optimization
- [x] Data science and statistics
- [x] Advanced data visualization (D3.js/Canvas)
- [x] IndexedDB and client-side data management
- [x] Web Workers and parallel processing
- [x] Mathematical modeling and algorithms
- [x] Time series analysis
- [x] Performance optimization

**Implementation Plan:**
- **Phase 1**: Design ML architecture and data collection system (5-6 days)
- **Phase 2**: Implement performance prediction models (6-7 days)
- **Phase 3**: Build pattern recognition and recommendation engine (5-6 days)
- **Phase 4**: Create advanced analytics dashboard (4-5 days)
- **Phase 5**: Add anomaly detection and trend analysis (3-4 days)
- **Phase 6**: Testing, model training, and optimization (3-4 days)

## 📚 Additional Context
**Use Cases/Scenarios:**
1. Research comparing algorithm efficiency across different data patterns
2. Teaching advanced algorithm analysis and selection principles
3. Performance engineering for large-scale applications
4. Algorithm optimization research and development
5. Predictive performance modeling for system design
6. Academic research on algorithm behavior patterns

**Advanced ML Features:**
- **Ensemble Methods**: Combine multiple ML models for better predictions
- **Online Learning**: Models that improve with each algorithm run
- **Transfer Learning**: Apply insights from one algorithm to others
- **Explainable AI**: Understand why specific recommendations are made
- **A/B Testing**: Automated algorithm comparison with statistical significance
- **Performance Profiling**: Deep analysis of bottlenecks and optimization opportunities

**Integration Points:**
- Current metrics system in `MetricsPanel.jsx`
- Performance tracking in `PerformanceMetrics.jsx`
- Historical data storage and retrieval
- Algorithm execution monitoring
- User interaction tracking for personalization

**Technical Challenges:**
- Training ML models with limited client-side data
- Real-time inference without blocking UI
- Balancing model complexity with browser performance
- Ensuring model accuracy across diverse use cases
- Managing large datasets in browser storage
- Cross-browser compatibility for TensorFlow.js

**Educational Benefits:**
- Introduces students to ML applications in computer science
- Demonstrates data-driven algorithm selection
- Teaches performance prediction and modeling
- Shows real-world application of ML in software engineering
- Develops understanding of statistical analysis in algorithms

## ✅ Checklist
- [x] I have searched existing issues to ensure this is not a duplicate
- [x] I have provided a clear problem statement and solution
- [x] I have considered alternative approaches
- [x] I have assessed the impact and priority
- [x] I have included relevant technical details

### PR DESCRIPTION
**Pull request by MRohit-2003**
# 🚀 Pull Request

⭐ Yes, I have starred the repository

## 📝 Description
This pull request solves Issue#43. It implies the first phase of the issue. It collects performance data from the algorithms and stores in an API in the backend.
MongoDB is used to store the database of algorithm performance data. A server is created in the backend using Express which collects the data from the user interaction and stores in the performance database .

**Related Issue(s):**
- [x] Fixes #43 
- [x] Closes #43 
- [x] Relates to #43 

## 🔄 Type of Change
Please delete options that are not relevant:

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 📚 Documentation update

### Core Changes
- [x]Added a data collection and storage server for the Issue #43 
- [x] Added a new "backend" folder
- [x] Added a new folder "models" in the backend folder
- [x] Created a new Performance.js file in the models folder
- [x] Updated the SortingVisualizer.js file

### Detailed Changes
1. **Change 1**: Added a code snippet in the SortingVisualizer to collect and store alogrithm performance data. It is named logPerformanceData
2. **Change 2**: Created a folder named "backend"
3. **Change 3**: Added a .env file to the backed folder
4. **Change 4**: initialised the backend folder
5. **Change 5**: Created a new "models" folder in the backend folder.
6. **Change 6**: Added a "Performance.js" file to the backed\models folder.

### Files Modified
- `C:\Users\mrohi\SortVision\SortVision\src\components\sortingVisualizer` - Added a code snippet to collect and store alogrithm performance data. It is named logPerformanceData

image

## 🏆 SSOC Season 4 (if applicable)
- [x] I'm participating in Social Summer of Code (SSOC) Season 4
- [x] This PR is part of my SSOC contribution

<details>
<summary>🎓 SSOC Details (Click to expand if participating)</summary>

### Contribution Information
**Issue Difficulty:** 
- [x] 🔴 Advanced (40 points)

**Estimated Time Spent:** : 1 week

**Skills Demonstrated:**
- [x] React/TypeScript development
- [ ] Algorithm implementation
- [ ] UI/UX design
- [ ] Performance optimization
- [x] Testing
- [x] Documentation
- [x] Accessibility

**Learning Outcomes:**
- How to connect a server to a database and store data
- Acquiring Sofware version updates with current project dependencies

</details>

### Code Quality
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or errors
- [x] I have added tests that prove my fix is effective or that my feature works

### Git Hygiene
- [x] My commits are atomic and have descriptive messages
- [x] I have rebased/merged the latest main branch
- [x] I have resolved any merge conflicts

### Project Specific
- [x] All sorting algorithms maintain their correctness
- [x] Visual representations are accurate
- [x] Performance is not degraded
- [x] Accessibility standards are maintained
- [x] Mobile responsiveness is preserved

## 👥 Reviewers

@alienx5499

---

**Note for Reviewers:**
- Focus on algorithm correctness and performance
- Check visual accuracy of sorting representations
- Verify responsive design and accessibility
- Test with different data sizes and edge cases

### CODE REVIEW

**Comment by alienx5499** [comment]
Hi @MRohit-2003 👋,

Thank you for your detailed PR and all the effort you've put into implementing the backend performance logging system!

Before proceeding, I noticed there are **merge conflicts** in the following files:

- `SortVision/package-lock.json`
- `SortVision/package.json`

Please resolve these so that I can proceed with the review and merge. ✅

That said — I want to share some thoughts on the feature. While your work is commendable and the first phase has been executed well, **after some thought, we feel this functionality may not be essential for SortVision at the moment**.

However, if you're still interested and excited to explore this idea further, I’d be happy to support it and we can shape it together going forward. Let me know your thoughts!

Regardless, for your hard work and clean execution, I’ll make sure to **award you the full SSOC points** and **close the issue** officially even if we don't take it live right away. 🚀

Thanks again for your initiative and contribution!

— Prabal Patra | SSOC Project Admin

**Comment by alienx5499** [comment]
Thank you once again @MRohit-2003 for the valuable contribution 🙌  
As discussed earlier, we’ve decided not to move forward with this feature right now — but your implementation is excellent.

I’m merging this for record and awarding you **full SSOC points** for Phase 1 🎯  
The feature will remain in the repo for future consideration, but won’t go live at the moment.

— Prabal Patra | SSOC Project Admin
